### ISSUE TITLE
IMOS: check_data_variables - XY coordinates defined but missing in var definition

### ISSUE THREAD

**Issue by lbesnard** [aodn/compliance-checker#18]
file: **/mnt/ebs/wip/ANMN/timeSeriesProfile_po_permission/new_tb/IMOS_ANMN-QLD_AETVZ_20130620T075900Z_ITFTIS_FV01_ITFTIS-1306-Workhorse-ADCP-464.5_END-20140111T055900Z_C-20150515T020357Z.nc**

**netcdf-checker**:

```
    data_variable                      :3:    94/112 :
        ABSI1                          :3:     4/ 5 :
            check_dimension            :3:     0/ 1 : dimension doesn't contain
                                                      TIME, LATITUDE, LONGITUDE,
                                                      DEPTH
```

**ncdump**:

```
        float ABSI1(TIME, HEIGHT_ABOVE_SENSOR) ;
                ABSI1:coordinates = "TIME LATITUDE LONGITUDE HEIGHT_ABOVE_SENSOR" ;
```

The IMOS checker seems to complain about missing XYZT coordinations from the variable's dimension definition. 

Looking at the code:

``` python
    def check_data_variables(self, ds):
        """
        Check data variable:
            at least one data variable exisits
        """
```

This function definition is ambiguous, not really as descriptive as CF ones. The more i look at the function, the less I understand the link between the function's definition and its code. What are we really talking about ?? data variable or list of dimensions in a data variable ? (plus typo "exisits"). If I let myself imagine that a data variable is in this case actually one of the variable's dimension (i.e. TIME and HEIGHT_ABOVE_SENSOR), then, this is not right since the **ABSI1** variable contains at least the__TIME__ coordinate.

In this case
1) the error message should be more specific and tell us which coordinate is missing
2) the function should not fail because TIME exists in the ABSI1 variable's definition, and the function definition says "at least one ..."

Let's see if I imagined it the right way
more python code, a bit deeper:

``` python
            required_dimensions = ['TIME', 'LATITUDE', 'LONGITUDE', 'DEPTH']

            for var in self._data_variables:
                result_name = ('var', 'data_variable', var.name, 'check_dimension')
                passed = True
                reasoning = None

                if len(var.dimensions) > 0:
                    for dimension in var.dimensions:
                        if dimension in required_dimensions:
                            passed = True
                        else:
                            passed = False
                else:
                    passed = False

                if not passed:
                    reasoning =  ["dimension doesn't contain TIME, LATITUDE, LONGITUDE, DEPTH"]
```

This part of the code makes me think that the function check_data_variables actually checks that the dimension of a variable contains dimensions from the `required_dimensions` python variable defined in the block above (so not at all like what suggested the function description's purpose). However, correct me if I'm wrong, but in the FOR loop within the `if len(var.dimensions) > 0` (see above) , it seems like the python variable `passed` is reset for each dimension. In our case, this means that for the ABSI1 variable,

**on iteration 1 )**
dimension = TIME
passed = True

**iteration 2)**
dimension = HEIGHT_ABOVE_SENSOR
passed = False

And then we get the message saying that the "dimension doesn't contain TIME ..." (FWIW, it's not the dimension which doesn't contain TIME, but the dimension list of the variable),  which  is contradictory to the function definition `at least one data variable exisits`  (not to mention again this sentence doesn't make sense to me in the context of this functions)

And if ABSI1 was defined as  `ABSI1(HEIGHT_ABOVE_SENSOR, TIME)` instead, I'm pretty sure the function would have passed this test.

As a summary, I'm really confused :
![Alt Text](http://www.commcreative.com/sites/default/files/field/image/Help!.JPG)

**Comment by mhidas** [comment]
@lbesnard I think the reason for the confusion is that @fxmzb123 implemented several checks in one method, because they all apply to data variables (i.e. any var that is not a coordinate or ancillary var - see checks definitions [here](https://github.com/mhidas/imos-netcdf-conventions/blob/master/IMOS_netCDF_checker_tests_for_manual_v1.3.xls)). This is ok. 

The check that's failing is that each data variable should have at least one of the dimensions time, lat, lon or depth.

I think you've correctly identified all the issues:
  1) The code comments don't explain everything this method is doing;
  2) The logic of the check is indeed wrong -- it should skip out of that for loop as soon as it finds one dimension that is in the `required_dimensions` list.
  3) The reasoning message is not quite clear. A more helpful message would be "Data variable should have at least one of the dimensions TIME, DEPTH, LATITUDE, LONGITUDE".

I don't think there's anything to fix the files.

**Comment by ggalibert** [comment]
Please see PR https://github.com/aodn/compliance-checker/pull/64

**Comment by mhidas** [comment]
Fixed via #64
